<a href="https://colab.research.google.com/github/Yatharth19/ML/blob/main/movie_recommender.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import ast# to convert string of list into a list
from sklearn.feature_extraction.text import CountVectorizer
from nltk.stem.porter import PorterStemmer
from sklearn.metrics.pairwise import cosine_similarity #to calculate cosine distance b/w every 2 movies
import pickle #to export this code data

In [ ]:
movies=pd.read_csv('/content/drive/MyDrive/movie_recommend/tmdb_5000_movies.csv')
credits=pd.read_csv('/content/drive/MyDrive/movie_recommend/tmdb_5000_credits.csv')

In [ ]:
movies.head()

budget  ... vote_count
0  237000000  ...      11800
1  300000000  ...       4500
2  245000000  ...       4466
3  250000000  ...       9106
4  260000000  ...       2124

[5 rows x 20 columns]

In [ ]:
credits.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 4 columns]

In [ ]:
credits['cast']

0       [{"cast_id": 242, "character": "Jake Sully", "...
1       [{"cast_id": 4, "character": "Captain Jack Spa...
2       [{"cast_id": 1, "character": "James Bond", "cr...
3       [{"cast_id": 2, "character": "Bruce Wayne / Ba...
4       [{"cast_id": 5, "character": "John Carter", "c...
                              ...                        
4798    [{"cast_id": 1, "character": "El Mariachi", "c...
4799    [{"cast_id": 1, "character": "Buzzy", "credit_...
4800    [{"cast_id": 8, "character": "Oliver O\u2019To...
4801    [{"cast_id": 3, "character": "Sam", "credit_id...
4802    [{"cast_id": 3, "character": "Herself", "credi...
Name: cast, Length: 4803, dtype: object

In [ ]:
credits['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [ ]:
credits.head(1)['cast'].values

array(['[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "ge

In [ ]:
movies.isnull().sum()

budget                     0
genres                     0
homepage                3091
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
dtype: int64

In [ ]:
credits.isnull().sum()

movie_id    0
title       0
cast        0
crew        0
dtype: int64

MERGING BOTH OF THESE DATAFRAMES BASED ON TITLE

In [ ]:
movies=movies.merge(credits,on='title')

In [ ]:
movies.shape

(4809, 23)

WE WOULD BE ONLY KEEPING GENRES, ID, KEYWORDS, TITLE, OVERVIEW, CAST, CREW

In [ ]:
movies=movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [ ]:
movies.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 7 columns]

In [ ]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [ ]:
movies.dropna(inplace=True)

NOW TO CHECK FOR ANY DUPLICATES

In [ ]:
movies.duplicated().sum()

0

In [ ]:
movies['genres'][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

NOW WE SHALL CONVERT THE FORMAT OF EACH GENRE AS FOLLOWS:



['ACTION','ADVENTURE','FANTASY','SCIENCE FICTION']

In [ ]:
def convert(obj):
  L=list()
  for i in ast.literal_eval(obj):
    L.append(i["name"])
  return L

NOW AS movies['genres'] CONTAINS A STRING OF LIST,WE NEED TO CONVERT IT INTO A LIST SO THAT NO ERROR IS RAISED

In [ ]:
movies['genres']=movies['genres'].apply(convert)

In [ ]:
movies.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 7 columns]

NOW WE DO THE SAME THING ON KEYWORDS COLUMN

In [ ]:
movies['keywords'][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [ ]:
movies['keywords']=movies['keywords'].apply(convert)

In [ ]:
movies.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 7 columns]

IN THE CAST COLUMN, WE NEED TO RETRIEVE THE TOP 3 ACTORS

In [ ]:
movies['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [ ]:
def convert3(obj):
  L=list()
  j=0
  for i in ast.literal_eval(obj):
    if j!=3:
      L.append(i["name"])
      j=j+1
    else:
      break
    
  return L

In [ ]:
movies['cast']=movies['cast'].apply(convert3)

In [ ]:
movies.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 7 columns]

In [ ]:
movies['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

NOW FROM THE CREW, WE ONLY WANT THE DIRECTOR'S NAME
SO WE SEARCH ALL DICTIONARIES AND TRY TO FIND WHERE THE JOB IS DIRECTOR

In [ ]:
def fetch_director(obj):
  L=list()
  j=0
  for i in ast.literal_eval(obj):
    if i["job"]=="Director":
      L.append(i['name'])
      return L
  

In [ ]:
movies['crew']=movies['crew'].apply(fetch_director)

In [ ]:
movies.head()

movie_id  ...                 crew
0     19995  ...      [James Cameron]
1       285  ...     [Gore Verbinski]
2    206647  ...         [Sam Mendes]
3     49026  ...  [Christopher Nolan]
4     49529  ...     [Andrew Stanton]

[5 rows x 7 columns]

In [ ]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

WE WOULD BE CONVERTING THE ABOVE STRING INTO A LIST SO THAT WE CAN MERGE IT WITH OTHER COLUMNS

In [ ]:
movies['overview']=movies['overview'].apply(lambda x: x.split())

In [ ]:
movies.head()

movie_id  ...                 crew
0     19995  ...      [James Cameron]
1       285  ...     [Gore Verbinski]
2    206647  ...         [Sam Mendes]
3     49026  ...  [Christopher Nolan]
4     49529  ...     [Andrew Stanton]

[5 rows x 7 columns]

NOW WE NEED TO REMOVE THE SPACES BETWEEN OUR TEXT AS OTHERWISE THE TAGS FORMED WOULD BE MISGUIDING...FOR EXAMPLE A B, A C.

In [ ]:
movies['crew'].isnull().sum()

30

In [ ]:
movies.dropna(inplace=True)

In [ ]:
movies['crew'].isnull().sum()

0

In [ ]:
movies['genres']=movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])
movies['keywords']=movies['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movies['cast']=movies['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movies['crew']=movies['crew'].apply(lambda x:[i.replace(" ","") for i in x])

In [ ]:
movies.head()

movie_id  ...                crew
0     19995  ...      [JamesCameron]
1       285  ...     [GoreVerbinski]
2    206647  ...         [SamMendes]
3     49026  ...  [ChristopherNolan]
4     49529  ...     [AndrewStanton]

[5 rows x 7 columns]

In [ ]:
movies['tags']=movies['overview']+movies['genres']+movies['keywords']+movies['cast']+movies['crew']

In [ ]:
movies.head()

movie_id  ...                                               tags
0     19995  ...  [In, the, 22nd, century,, a, paraplegic, Marin...
1       285  ...  [Captain, Barbossa,, long, believed, to, be, d...
2    206647  ...  [A, cryptic, message, from, Bond’s, past, send...
3     49026  ...  [Following, the, death, of, District, Attorney...
4     49529  ...  [John, Carter, is, a, war-weary,, former, mili...

[5 rows x 8 columns]

In [ ]:
new_df=movies[['movie_id','title','tags']]

In [ ]:
new_df.head()

movie_id  ...                                               tags
0     19995  ...  [In, the, 22nd, century,, a, paraplegic, Marin...
1       285  ...  [Captain, Barbossa,, long, believed, to, be, d...
2    206647  ...  [A, cryptic, message, from, Bond’s, past, send...
3     49026  ...  [Following, the, death, of, District, Attorney...
4     49529  ...  [John, Carter, is, a, war-weary,, former, mili...

[5 rows x 3 columns]

NOW WE CONVERT THE ABOVE LIST IN A STRING

In [ ]:
new_df['tags']=new_df['tags'].apply(lambda x:" ".join(x))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
new_df.head()

movie_id  ...                                               tags
0     19995  ...  In the 22nd century, a paraplegic Marine is di...
1       285  ...  Captain Barbossa, long believed to be dead, ha...
2    206647  ...  A cryptic message from Bond’s past sends him o...
3     49026  ...  Following the death of District Attorney Harve...
4     49529  ...  John Carter is a war-weary, former military ca...

[5 rows x 3 columns]

NOW WE CONVERT INTO LOWERCASE AS IT IS RECOMMENDED


In [ ]:
new_df['tags']=new_df['tags'].apply(lambda x:x.lower())

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
new_df.head()

movie_id  ...                                               tags
0     19995  ...  in the 22nd century, a paraplegic marine is di...
1       285  ...  captain barbossa, long believed to be dead, ha...
2    206647  ...  a cryptic message from bond’s past sends him o...
3     49026  ...  following the death of district attorney harve...
4     49529  ...  john carter is a war-weary, former military ca...

[5 rows x 3 columns]

NOW WE DO TEXT VECTORIZATION..... WE CONVERT OUR TEXT TO VECTORS AND THEN RECOMMEND THE CLOSEST VECTOR TO THE CONSUMER

WE ALSO WON'T BE USING ANY STOPWORDS IN VECTORIZATION.
STOP WORDS ARE WORDS LIKE IN, ARE, THE ,A ETC.

NOW WE WILL BE CONVERTING ALL WORDS IN THEIR ROOT FORM.....USE PORTER STEMMER (USING NLTK)

In [ ]:
ps=PorterStemmer()

In [ ]:
def stem(text):
  y=[]

  for i in text.split():
    y.append(ps.stem(i))

  return " ".join(y)


In [ ]:
new_df['tags']=new_df['tags'].apply(stem)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
cv=CountVectorizer(max_features=5000,stop_words="english")


In [ ]:
vectors=cv.fit_transform(new_df['tags']).toarray()

In [ ]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0])

In [ ]:
cv.get_feature_names()#this will return the 5000 most frequently used words

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['000',
 '007',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '17th',
 '18',
 '18th',
 '18thcenturi',
 '19',
 '1910',
 '1920',
 '1930',
 '1940',
 '1944',
 '1950',
 '1950s',
 '1960',
 '1960s',
 '1970',
 '1970s',
 '1974',
 '1976',
 '1980',
 '1985',
 '1990',
 '1999',
 '19th',
 '19thcenturi',
 '20',
 '200',
 '2003',
 '2009',
 '20th',
 '21st',
 '23',
 '24',
 '25',
 '30',
 '300',
 '3d',
 '40',
 '50',
 '500',
 '60',
 '70',
 '80',
 'aaron',
 'aaroneckhart',
 'abandon',
 'abduct',
 'abigailbreslin',
 'abil',
 'abl',
 'aboard',
 'abov',
 'abus',
 'academ',
 'academi',
 'accept',
 'access',
 'accid',
 'accident',
 'acclaim',
 'accompani',
 'accomplish',
 'account',
 'accus',
 'ace',
 'achiev',
 'acquaint',
 'act',
 'action',
 'actionhero',
 'activ',
 'activist',
 'activities',
 'actor',
 'actress',
 'actual',
 'adam',
 'adamsandl',
 'adamshankman',
 'adapt',
 'add',
 'addict',
 'adjust',
 'admir',
 'admit',
 'adolesc',
 'adopt',
 'ador',
 'adrienbrodi',
 'adult',
 'adultanim',


NOW WE CALCULATE THE COSINE DISTANCE BETWEEN PAIR OF EVERY 2 VECTORS.....NOT EUCLIDEAN DISTANCE AS IT IS NOT GOOD AT THIS POINT WITH HIGH DIMENSIONALITY

In [ ]:
cosine_similarity(vectors).shape

(4776, 4776)

IT IS BECAUSE FOR 1 MOVIE, WE HAVE TO FIND DISTANCE FROM 4776 MOVIES

In [ ]:
similarity=cosine_similarity(vectors)

In [ ]:
similarity[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

In [ ]:
similarity[1]

array([0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
       0.02615329])

In [ ]:
def recommend(movie):
  movie_index=new_df[new_df['title']==movie].index[0]
  distances=similarity[movie_index]
  movies_list=sorted(list(enumerate(distances)),reverse=True,key=lambda x:x[1])[1:6]

  for i in movies_list:
    print(new_df.iloc[i[0]].title)

In [ ]:
recommend('Aliens') 

Alien
Alien³
Meet Dave
Star Trek: Insurrection
Escape from Planet Earth


In [ ]:
recommend('Titanic')

The Notebook
Under the Same Moon
Ghost Ship
The Bounty
Pirates of the Caribbean: On Stranger Tides


In [ ]:
new_df.head()

movie_id  ...                                               tags
0     19995  ...  in the 22nd century, a parapleg marin is dispa...
1       285  ...  captain barbossa, long believ to be dead, ha c...
2    206647  ...  a cryptic messag from bond’ past send him on a...
3     49026  ...  follow the death of district attorney harvey d...
4     49529  ...  john carter is a war-weary, former militari ca...

[5 rows x 3 columns]

In [ ]:
pickle.dump(new_df,open('movies.pkl','wb'))

In [ ]:
new_df['title'].values

array(['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre',
       ..., 'Signed, Sealed, Delivered', 'Shanghai Calling',
       'My Date with Drew'], dtype=object)

In [ ]:
pickle.dump(similarity,open('similarity.pkl','wb'))